validator/tests/unit3/test_scheduler/tests.py

```diff
@@ -299,104 +297,788 @@ def test_valid_batch_invalid_batch(self):
+        '''
+
+        self.show_tree()
+
+        for address, rwc in expected_dict.items():
+            readers, writer, children = rwc
+            node = self.get_node(address)
+            self.assertIsNotNone(node)
+            self.assertEqual(readers, node.readers)
+            self.assertEqual(writer, node.writer)
+            self.assertEqual(
+                set(children),
+                set(node.children.keys()))
+
+    def assert_no_nodes_at_addresses(self, *addresses):
+        for address in addresses:
+            node = self.get_node(address)
+            self.assertIsNone(node)
+
+    def assert_rw_count(self, reader_count, writer_count):
+        '''
+        Asserts the total number of readers and writers in the tree
+        '''
+
+        def count_readers(node=None):
+            if node is None:
+                node = self.get_node('') # root node
+            count = len(node.readers)
+            for child in node.children:
+                next_node = node.children[child]
+                count += count_readers(next_node)
+            return count
+
+        def count_writers(node=None):
+            if node is None:
+                node = self.get_node('') # root node
+            count = 1 if node.writer is not None else 0
+            for child in node.children:
+                next_node = node.children[child]
+                count += count_writers(next_node)
+            return count
+
+        self.assertEqual(reader_count, count_readers())
+        self.assertEqual(writer_count, count_writers())
+
+    def assert_rw_preds_at_addresses(self, rw_pred_dict):
+        '''
+        Asserts the read predecessors and write predecessors at aan address
+
+        rw_pred_dict = {address: (read_preds, write_preds)}
+        '''
+
+        for address, rw_preds in rw_pred_dict.items():
+            read_preds, write_preds = rw_preds
+
+            self.assertEqual(
+                self.tree.find_read_predecessors(address),
+                set(read_preds))
+
+            self.assertEqual(
+                self.tree.find_write_predecessors(address),
+                set(write_preds))
+
+    # basic tree operations (for convenience)
+
+    def add_readers(self, reader_dict):
+        '''
+        reader_dict = {address: reader}
+        '''
+        for address, reader in reader_dict.items():
+            self.add_reader(address, reader)
+
+    def set_writers(self, writer_dict):
+        '''
+        writer_dict = {address: writer}
+
+        WARNING: The order in which writers are set is not specified.
+        '''
+        for address, writer in writer_dict.items():
+            self.set_writer(address, writer)
+
+    def add_reader(self, address, txn):
+        self.tree.add_reader(address, txn)
+
+    def set_writer(self, address, txn):
+        self.tree.set_writer(address, txn)
+
+    def get_node(self, address):
+        return self.tree.get(address)
+
+    # display
+
+    def show_tree(self):
```

Comment:
Because of the newlines here, the approach would be to split on newline and log each line separately; ideally, prefixing the line with something friendly to grep.